### ISSUE TITLE
Check/fix 'WebGLRenderer' component's modes 'always'  & 'auto' + interaction

### ISSUE THREAD

**Issue by vatro** [vatro/svelthree#63]

**Comment by vatro** [comment]
- [x] "on every frame" = mode: "always" -> [1.0.0-next.0.45](https://github.com/vatro/svelthree/tree/1.0.0-next.0.45)
- [x] "once" = mode "once" -> [1.0.0-next.0.45](https://github.com/vatro/svelthree/tree/1.0.0-next.0.45)
"on change" = mode "change" -> todo: #64 

Mode "change" has not been implemented yet, leaving it for `next.2` release.

**Comment by vatro** [comment]
The `mode:"always"` (new approach) performance is bad in scenes with a lot of interactive objects (_noticed in the GLTF-Rover example_) ~beacuse all interactive objects are being raycasted on every frame (_previously: on pointermove_)~ ...

**Comment by vatro** [comment]
The problem was actually bigger, the 'raycast' array filtering was wrong. While fixing it, I've also almost completely refactored interaction functionality with some new features:
- added more pointer events
- added 'block' and 'change_cursor' props
- added 'on' and 'onx' functions
- various performance optimizations 
etc.

see [1.0.0-next.0.52](https://github.com/vatro/svelthree/tree/1.0.0-next.0.52)
see [1.0.0-next.0.53](https://github.com/vatro/svelthree/tree/1.0.0-next.0.53)

**Comment by vatro** [comment]
While trying to fix / optimize **mode: `"once"`**, I came to the conclusion that it's **obsolete**. The _render only once_ functionality  can be merged with the `"auto"`-mode functionality which should render once + when something changes. **Changes** triggering a re-render can be modified or replaced geometries, transformed objects, modified materials, interactivity configuration etc. **important!** -> **applied via components' API / references / accessors** **(_must trigger a component update_)**.

In `"once"`-mode a re-render would have to be triggered manually by the user every time something changes (_even if this was not intially intended_), which is cumbersome (_bad dx_) and can quickly lead to ubercomplex (_not needed / unwanted_) logic + is especially unintuitive when dealing with interactivity changes (_scene's appearance doesn't change_) for which we would need a new public `WebGLRenderer` / `Canvas` (?) function (e.g. _`update_interactivity()`_) and additional logic in most components in order to bypass re-rendering and update the interactivity state of objects and the `Canvas` component only (_without rerendering_). Currently, interactivity updates are handled before the actual render inside the `WebGLRenderer` components's `render_standard()` function.

**short**:
`WebGLRenderer` **mode `"once"`** concept is **deprecated**, instead we'll have only `"always"` and `"auto"` as _'render once' on steroids_.

**Comment by vatro** [comment]
ok, mode `"auto"` is implemented -> #64 

I've already tested some crucial test-scenes, but it would we best to have each test-scene in both modes `"always"` and `"auto"` as there are some important differences on how / when to (_best_) apply changes to the scene -> e.g. `on:before_render={...}` with mode `"always"` vs. `on:after_render={...}` + / or `setTimeout(...)`, `requestAnimationFrame(...)` etc. with mode `"auto"`.

Also, some ideas for new `WebGLRenderer` functions came to my mind, like: `update(?)(start step / frame (optional) : number)` for immediate rendering (_at any point_) with a possibility to set `WebGLRenderer` component's internal frame counter to some start value / some value -> the frame counter / next frame (number) could then again be used via `on:before_render` or `on:after_render` for some interesting behavior.

**Comment by vatro** [comment]
I've refactored the `WebGLRenderer` component and did some extensive polishing on both modes over the last 2-3 months, resulting in the latest major refactor: f20757f4353bb9cdb3cafccac0e54c68e99f04b4.

Both modes (`"auto"` and `"always"`) feel pretty robust by now. Meanwhile I started to like mode "`auto`" more, because it has some cool benefits over `"always"` like performance (_of course, if nothing changes, the scene doesn't have to be re-rendered -> idle_) and event handling (_events are more direct, not queued like in mode `"always"`, which allows using e.g. `ev.preventDefault()` inside event handlers and also results in a true reactive rendering if something is changed via event handlers._), so I've made **`"auto"`** the **new default render mode**. I even started thinking about maybe dropping the `"always"`mode completely at some point in future, we shall see.

I'm closing this issue because it's too general, if needed, I'll open some new, more specific ones concerning both render modes.
